File: go/backtracking/22_generate_parenthesis.go
package backtracking

import "strings"

// generateParenthesis 括号生成
func generateParenthesis(n int) []string {
	// 回溯搜索

	// 所有满足条件的集合
	res := make([]string, 0)

	// 搜索中满足条件的集合
	path := make([]string, 0)

	// symbol 记录当前左括号的个数, symbol < n
	// path 中push ( symbol + 1; push ） symbol - 1, 最后symbol 被抵消==0
	// idx path 中当前元素下标
	var backtracck func(symbol, idx int)
	backtracck = func(symbol, idx int) {
		if idx == n*2 {
			if symbol == 0 {
				p := strings.Join(path, "")
				res = append(res, p)
			}
			return
		}

		// 可以往path栈里面push (, 每次push 左括号 symbol + 1
		if symbol < n {
			path = append(path, "(")
			backtracck(symbol+1, idx+1)
			path = path[:len(path)-1]
		}
		// 往path 里面push ), 抵消左括号数量
		if symbol > 0 {
			path = append(path, ")")
			backtracck(symbol-1, idx+1)
			path = path[:len(path)-1]
		}
	}

	backtracck(0, 0)
	return res
}

File: rust/src/backtracking/_22_generate_parenthesis.rs
/// 括号生成
pub fn generate_parenthesis(n: i32) -> Vec<String> {
    // 满足条件的所有结果的合集
    let mut res = vec![];
    // 当前搜索过程中满足条件的路径
    let mut path = vec![];

    let n = n as usize;

    // symbol 当前左括号的个数，symbol < n , push (; if push ), symbol -1
    fn backtrack(symbol: usize, idx: usize, n: usize, path: &mut Vec<&str>, res: &mut Vec<String>) {
        if idx == n * 2 {
            if symbol == 0 {
                res.push(path.join(""));
            }
            return;
        }

        // symbol<n， 则可以push(
        if symbol < n {
            path.push("(");
            backtrack(symbol + 1, idx + 1, n, path, res);
            // backtrack
            path.pop();
        }

        if symbol > 0 {
            path.push(")");
            backtrack(symbol - 1, idx + 1, n, path, res);
            path.pop();
        }
    }

    backtrack(0, 0, n, &mut path, &mut res);

    res
}
